File: src/id0070.c
// Licensed under the MIT License.

// Totient Permutation

#include <math.h>
#include "../lib/euler.h"
#include "../lib/string_builder.h"
#include "../lib/permutation_iterator.h"
#include "../lib/sieve_iterator.h"
#define MAX_SEARCH 0.375

int main(void)
{
    double mid = sqrt(1e7);
    double minNRPhi = INFINITY;
    long minN = -1;
    int min = mid * (1 - MAX_SEARCH);
    int max = mid * (1 + MAX_SEARCH);
    struct Sieve primes;
    struct SieveIterator p;
    struct StringBuilder nDigits;
    struct StringBuilder phiDigits;
    clock_t start = clock();

    euler_ok(sieve(&primes, max));
    euler_ok(string_builder(&nDigits, 0));
    euler_ok(string_builder(&phiDigits, 0));
    sieve_begin(&p, &primes);

    for (sieve_jump(&p, min); p.current < max; sieve_next(&p))
    {
        struct SieveIterator q;

        for (q = p; ; sieve_next(&q))
        {
            long n = p.current * q.current;

            if (n > 10000000l)
            {
                break;
            }

            long phi = n - p.current - q.current + 1;
            double nRPhi = (double)n / phi;

            if (nRPhi >= minNRPhi)
            {
                continue;
            }

            string_builder_clear(&nDigits);
            euler_ok(string_builder_append_format(&nDigits, "%ld", n));
            string_builder_clear(&phiDigits);
            euler_ok(string_builder_append_format(&phiDigits, "%ld", phi));

            if (permutation_test(
                nDigits.buffer,
                nDigits.length,
                phiDigits.buffer,
                phiDigits.length,
                1,
                char_equality_comparer))
            {
                minN = n;
                minNRPhi = nRPhi;
            }
        }
    }

    finalize_sieve(&primes);
    finalize_string_builder(&nDigits);
    finalize_string_builder(&phiDigits);
    
    return euler_submit(70, minN, start);
}

File: src/id0070.py
# Licensed under the MIT License.

# Totient Permutation

from math import inf, sqrt
from sympy import primerange
from time import time

MAX_SEARCH = 0.375

minN = -1
minNRPhi = inf
minPrime = int(sqrt(1e7) * (1 - MAX_SEARCH))
maxPrime = int(sqrt(1e7) * (1 + MAX_SEARCH))
start = time()

for p in primerange(minPrime, maxPrime):
    for q in primerange(p):
        n = p * q
        
        if n > 1e7: break
        
        phi = (p - 1) * (q - 1)
        nRPhi = n / phi
        
        if nRPhi >= minNRPhi or sorted(str(n)) != sorted(str(phi)): continue
        
        minN = n
        minNRPhi = nRPhi

print(f"0070{minN:>64}    {time() - start:.6f}")
